keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1215 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Introduce new set of APIs to express an administrator's intent in securing
+their K8s cluster. This doc proposes the AdminNetworkPolicy API to complement
+the developer focused NetworkPolicy API in Kubernetes.
+
+## Motivation
+
+Kubernetes provides the NetworkPolicy resource to control traffic within a
+cluster. NetworkPolicy focuses on expressing a developer's intent to secure
+their applications. However, it was not intended to be used for cluster scoped
+administrative traffic control, which is reflected by its design:
+- NetworkPolicy uses a "implicit isolation" model, which means that once a policy
+  is applied to certain workloads, they are automatically isolated (in the direction
+  specified by the policy) and anything allowed needs to be explicitly called out.
+  It has no concept of explicit "deny" rules, because the application deployer can
+  simply refrain from allowing the things they want to deny.
+- It doesn't include priorities, or any concept of one user "outranking" another
+  and being able to override their policies, because a given application is expected
+  to only have policies created by a single user (the one deploying the application).
+Thus, in order to satisfy the needs of a cluster admin, we propose to introduce
+a new API that captures the administrator's intent.
+
+### Goals
+
+The goals for this KEP are to satisfy the following key user stories:
+
+1. As a cluster administrator, I want to enforce irrevocable in-cluster guardrails 
+   that all workloads must adhere to in order to guarantee the safety of my clusters.
+   In particular I want to enforce certain network level access controls that are
+   cluster scoped and cannot be overridden or bypassed by namespace scoped
+   NetworkPolicies.
+
+   Example: I would like to explicitly allow all pods in my cluster to reach
```

Comment:
This example is backwards, I think (and belongs in number 3), given that standard NP can only express Allow intent. Perhaps a better one might be:

Example: I would like to ensure that no pods are able to access my cloud provider's metadata server.

Reply:
Ah, scratch that - I didn't see that this KEP explicitly ignores external peers.

Reply:
This is a minor wording issue, but I personally associate "guardrail" in the context of ACLs with "blocking". Perhaps "invariant" is a better term?